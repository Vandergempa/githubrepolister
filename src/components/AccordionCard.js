import React from "react"

import Accordion from 'react-bootstrap/Accordion'
import Card from 'react-bootstrap/Card'

const AccordionCard = (props) => {
  return (
    <Card>
      <Accordion.Toggle as={Card.Header} eventKey={props.eventKey}>
        <strong>{props.title}</strong>
      </Accordion.Toggle>
      <Accordion.Collapse eventKey={props.eventKey}>
        <Card.Body>{props.body}</Card.Body>
      </Accordion.Collapse>
    </Card>
  )
}

export default AccordionCard