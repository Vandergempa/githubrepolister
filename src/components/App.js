import React from 'react';
import '../styles/App.css'
import 'bootstrap/dist/css/bootstrap.min.css';
import AppRouter from '../routers/AppRouter'

const App = () => {
  return (
    <AppRouter />
  );
}

export default App;
