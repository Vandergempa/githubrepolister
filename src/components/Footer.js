import React from "react"

const Footer = () => {
  return (
    <footer className="pl-5 pr-5 d-flex" style={{ height: "150px" }}>
      <p className="m-auto pt-3 w-100 text-center border-top">Â© 2019 Made By Tamas Bogdan</p>
    </footer>
  )
}

export default Footer