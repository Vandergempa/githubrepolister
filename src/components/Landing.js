import React, { Fragment } from "react"
import { withRouter } from 'react-router-dom'

import Button from 'react-bootstrap/Button'
import InputGroup from 'react-bootstrap/InputGroup'
import FormControl from 'react-bootstrap/FormControl'
import Col from 'react-bootstrap/Col'
import Modal from 'react-bootstrap/Modal'

export const Landing = (props) => {
  const [userName, setUsername] = React.useState(props.username);
  const [show, setShow] = React.useState(false);

  const renderModal = () => {
    return (
      <Fragment>
        <Modal show={show} onHide={() => setShow(false)} className="d-flex align-items-center justify-content-center">
          <Modal.Header closeButton>
            <Modal.Title>Ooopps</Modal.Title>
          </Modal.Header>
          <Modal.Body>You must enter a valid username!</Modal.Body>
        </Modal>
      </Fragment>
    )
  }

  const handleInputChange = (e) => {
    setUsername(e.target.value)
  }

  const handleButtonClick = () => {
    return fetch(`https://api.github.com/users/${userName}/repos`, {
      method: 'get',
      headers: { 'Content-Type': 'application/json' }
    }).then(res => {
      if (res.ok) {
        return res.json();
      } else {
        setShow(true)
        throw Error(`Request rejected with status ${res.status}`);
      }
    }).then(repos => {
      if (!userName) {
        setShow(true)
      } else if (userName && repos !== undefined) {
        props.history.push({
          pathname: '/reposlist',
          state: { repos: repos, username: userName }
        })
      }
    })
      .catch(e => console.log(e))
  }


  return (
    <section className="d-flex flex-column bgimage position-fixed">
      <div className="d-flex flex-column mt-auto mb-5">
        {renderModal()}
        <h1 className="text-center text-white display-3 custfontfamily">Github repo lister</h1>
        <InputGroup className="justify-content-center mt-5" size="lg">
          <InputGroup.Prepend>
            <InputGroup.Text id="btnGroupAddon">@</InputGroup.Text>
          </InputGroup.Prepend>
          <Col xs="auto" className="pl-0">
            <FormControl
              type="text"
              placeholder="Type Github username"
              aria-label="Type Github username"
              aria-describedby="btnGroupAddon"
              onChange={handleInputChange}
              size="lg"
            />
          </Col>
          <Button
            className="justify-content-center shadow-none ml-2 mr-3"
            onClick={handleButtonClick}
            size="lg"
            variant="light">
            Find & List
            </Button>
        </InputGroup>
      </div>
    </section>
  )
};

export default withRouter(Landing);
