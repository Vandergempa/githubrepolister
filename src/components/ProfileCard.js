import React from "react"

import Card from 'react-bootstrap/Card'
import Button from 'react-bootstrap/Button'

const ProfileCard = (props) => {
  const { userInfo } = props
  return (
    <Card className="col-xs-5 col-sm-6 m-5 text-center mb-auto">
      <Card.Img className="rounded-circle ml-auto mr-auto mt-2" variant="top" src={userInfo.avatar_url} style={{ width: '13rem' }} />
      <Card.Body>
        <Card.Title>{userInfo.name}</Card.Title>
        <Card.Text>
          <em>{userInfo.bio}</em>
        </Card.Text>
        <hr></hr>
        <Card.Text>
          Public repos: {userInfo.public_repos}
        </Card.Text>
        <Card.Text>
          Followers: {userInfo.followers} | Following: {userInfo.following}
        </Card.Text>
        <Card.Text>
          Created at: {userInfo.created_at}
        </Card.Text>
        <Button variant="primary" href={userInfo.html_url} >Go to profile</Button>
      </Card.Body>
    </Card>
  )
}

export default ProfileCard