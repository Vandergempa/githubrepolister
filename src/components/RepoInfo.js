import React, { Fragment } from "react"
import AccordionCard from './AccordionCard'

import Navbar from 'react-bootstrap/Navbar'
import Breadcrumb from 'react-bootstrap/Breadcrumb'
import Accordion from 'react-bootstrap/Accordion'

const RepoInfo = (props) => {
  const repoDetails = props.location.state.details
  return (
    <Fragment>
      <header>
        <Navbar bg="light" expand="sm">
          <img className="github-icon mr-2" src="https://unicons.iconscout.com/release/v2.0.1/svg/line/github.svg" alt="Kiwi standing on oval" />
          <Navbar.Brand href="/">GitHub Repo Lister</Navbar.Brand>
          <Navbar.Toggle aria-controls="basic-navbar-nav" />
          <Navbar.Collapse className="justify-content-end">
            <Navbar.Text className="cursor-pointer" onClick={() => window.history.back()}>
              Back
            </Navbar.Text>
          </Navbar.Collapse>
        </Navbar>
        <Breadcrumb bg="light">
          <Breadcrumb.Item active>Main</Breadcrumb.Item>
          <Breadcrumb.Item active>{repoDetails.owner.login}</Breadcrumb.Item>
          <Breadcrumb.Item active>{repoDetails.name}</Breadcrumb.Item>
        </Breadcrumb>
      </header>
      <section className="w-100">
        <h1 className="text-center mb-5 mt-5 display-3 custfontfamily">{repoDetails.name}</h1>
        <Accordion className="mb-2 cursor-pointer col-6 mr-auto ml-auto" defaultActiveKey="0">
          <AccordionCard eventKey="0" title="Description" body={repoDetails.description} />
          <AccordionCard eventKey="1" title="Repo URL" body={<a href={repoDetails.html_url}>{repoDetails.html_url}</a>} />
          <AccordionCard eventKey="2" title="Homepage" body={<a href={repoDetails.homepage}>{repoDetails.homepage}</a>} />
          <AccordionCard eventKey="3" title="Created At" body={repoDetails.created_at} />
          <AccordionCard eventKey="4" title="Updated At" body={repoDetails.updated_at} />
          <AccordionCard eventKey="5" title="Clone URL - SSH" body={repoDetails.ssh_url} />
          <AccordionCard eventKey="6" title="Main language" body={repoDetails.language} />
          <AccordionCard eventKey="7" title="Fork count" body={repoDetails.forks} />
          <AccordionCard eventKey="8" title="Watchers count" body={repoDetails.watchers} />
        </Accordion>
      </section>
    </Fragment>
  )
}

export default RepoInfo