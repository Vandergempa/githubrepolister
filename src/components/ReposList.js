import React, { Fragment } from "react"
import { withRouter } from 'react-router-dom'
import ProfileCard from './ProfileCard'

import ListGroup from 'react-bootstrap/ListGroup'
import Navbar from 'react-bootstrap/Navbar'
import Form from 'react-bootstrap/Form'
import FormControl from 'react-bootstrap/FormControl'
import Breadcrumb from 'react-bootstrap/Breadcrumb'

export const ReposList = (props) => {
  const repos = props.location.state.repos
  const userName = props.location.state.username
  // We do this for testing purposes only:
  let fetchUserInfo = props.location.state

  const [reposList, setReposList] = React.useState(repos);
  const [userInfo, setUserInfo] = React.useState({});

  fetchUserInfo = () => {
    return fetch(`https://api.github.com/users/${userName}`, {
      method: 'get',
      headers: { 'Content-Type': 'application/json' }
    }).then(res => {
      if (res.ok) {
        return res.json();
      } else {
        throw Error(`Request rejected with status ${res.status}`);
      }
    })
      .catch(e => console.log(e))
  }

  React.useEffect(() => {
    fetchUserInfo().then((user) => setUserInfo(user))
  }, [])

  const handleInputChange = (e) => {
    const newReposList = repos.filter((repo) => {
      return repo.name.toLowerCase().includes(e.target.value.toLowerCase())
    })
    setReposList(newReposList)
  }

  return (
    <Fragment>
      <header>
        <Navbar bg="light" expand="sm">
          <img className="github-icon mr-2" src="https://unicons.iconscout.com/release/v2.0.1/svg/line/github.svg" alt="Kiwi standing on oval" />
          <Navbar.Brand href="/">GitHub Repo Lister</Navbar.Brand>
          <Navbar.Toggle aria-controls="basic-navbar-nav" />
          <Navbar.Collapse id="basic-navbar-nav">
            <Form inline className="ml-auto">
              <FormControl
                type="text"
                placeholder="Search repositories..."
                className="mr-sm-2"
                onChange={handleInputChange}
              />
            </Form>
          </Navbar.Collapse>
        </Navbar>
        <Breadcrumb bg="light">
          <Breadcrumb.Item active>Main</Breadcrumb.Item>
          <Breadcrumb.Item active>{userInfo.login}</Breadcrumb.Item>
        </Breadcrumb>
      </header>
      <section className="h-100 d-flex flex-column justify-content-center align-items-center mt-5 mb-5">
        <h1 className="text-center display-3 custfontfamily">{userInfo.login}'s repositories</h1>
        <div className="h-100 d-flex flex-row justify-content-center align-items-center flex-wrap">
          <ProfileCard userInfo={userInfo} />
          {reposList.length !== 0 ? <ListGroup className="col-xs-auto mt-5">
            {reposList.map((repo, index) => {
              return <ListGroup.Item
                key={index}
                action
                onClick={() => props.history.push({
                  pathname: '/repoinfo',
                  state: { details: repo }
                })}
              >
                {repo.name}
              </ListGroup.Item>
            })}
          </ListGroup> : <h5 className="mt-5"><em>No repositories found...</em></h5>}
        </div>
      </section>
    </Fragment>
  )
};

export default withRouter(ReposList);