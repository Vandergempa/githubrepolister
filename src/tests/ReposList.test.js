import React from 'react';
import { shallow } from 'enzyme';
import { ReposList } from '../components/ReposList';

let wrapper

// This is a new way of dealing with React Hooks:
let setState = jest.fn()
let useStateSpy = jest.spyOn(React, 'useState')
useStateSpy.mockImplementation((init) => [init, setState]);

// We mock the fetch call:
const mockSuccessResponse = {};
const mockJsonPromise = Promise.resolve(mockSuccessResponse);
const mockFetchPromise = Promise.resolve({
  json: () => mockJsonPromise,
});
jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

beforeEach(() => {
  jest.spyOn(React, 'useEffect').mockImplementation(f => f());

  const props = {
    location: {
      state: {
        repos: [
          {
            name: "app#1"
          },
          {
            name: "app#2"
          },
          {
            name: "app#3"
          }
        ],
        username: "Tomi",
        fetchUserInfo: jest.fn().mockResolvedValue({})
      }
    }
  }

  wrapper = shallow(
    <ReposList
      {...props}
    />
  );
});

afterEach(() => {
  jest.clearAllMocks();
});


test('Should render the ReposList component correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should call fetch on component mount', () => {
  expect(global.fetch).toHaveBeenCalledWith(`https://api.github.com/users/Tomi`, {
    method: 'get',
    headers: { 'Content-Type': 'application/json' }
  })
});

test('should call setState after fetch', () => {
  expect(setState).toBeCalled()
});

test('should filter the repos on form change', () => {
  const FormControl = wrapper.find("FormControl")
  expect(FormControl.length).toEqual(1)

  FormControl.prop('onChange')({
    target: {
      value: 'app#2'
    }
  });

  expect(setState).toHaveBeenCalledWith([
    {
      name: "app#2"
    }
  ])
});

