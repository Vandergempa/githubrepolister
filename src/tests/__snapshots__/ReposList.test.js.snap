// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the ReposList component correctly 1`] = `
<Fragment>
  <header>
    <Navbar
      bg="light"
      collapseOnSelect={false}
      expand="sm"
      variant="light"
    >
      <img
        alt="Kiwi standing on oval"
        className="github-icon mr-2"
        src="https://unicons.iconscout.com/release/v2.0.1/svg/line/github.svg"
      />
      <NavbarBrand
        href="/"
      >
        GitHub Repo Lister
      </NavbarBrand>
      <NavbarToggle
        aria-controls="basic-navbar-nav"
        label="Toggle navigation"
      />
      <NavbarCollapse
        id="basic-navbar-nav"
      >
        <Form
          className="ml-auto"
          inline={true}
        >
          <FormControl
            className="mr-sm-2"
            onChange={[Function]}
            placeholder="Search repositories..."
            type="text"
          />
        </Form>
      </NavbarCollapse>
    </Navbar>
    <Breadcrumb
      bg="light"
      label="breadcrumb"
      listProps={Object {}}
    >
      <BreadcrumbItem
        active={true}
      >
        Main
      </BreadcrumbItem>
      <BreadcrumbItem
        active={true}
      />
    </Breadcrumb>
  </header>
  <section
    className="h-100 d-flex flex-column justify-content-center align-items-center mt-5 mb-5"
  >
    <h1
      className="text-center display-3 custfontfamily"
    >
      's repositories
    </h1>
    <div
      className="h-100 d-flex flex-row justify-content-center align-items-center flex-wrap"
    >
      <ProfileCard
        userInfo={Object {}}
      />
      <ListGroup
        className="col-xs-auto mt-5"
        horizontal={null}
        variant={null}
      >
        <ListGroupItem
          action={true}
          active={false}
          disabled={false}
          key="0"
          onClick={[Function]}
          variant={null}
        >
          app#1
        </ListGroupItem>
        <ListGroupItem
          action={true}
          active={false}
          disabled={false}
          key="1"
          onClick={[Function]}
          variant={null}
        >
          app#2
        </ListGroupItem>
        <ListGroupItem
          action={true}
          active={false}
          disabled={false}
          key="2"
          onClick={[Function]}
          variant={null}
        >
          app#3
        </ListGroupItem>
      </ListGroup>
    </div>
  </section>
</Fragment>
`;
